### PR TITLE
SR-4: homework 4

### PR DESCRIPTION
**Pull request by Sofiiar**
***Description***
- removed unused import from homework_3
- added homework_4 with the following structure:
      - conftest -> fixture 
      - constants -> repetitive values
      - locators -> used xpath as a standard
      - test_demoqa -> 2 test scenarious: `test_alert_confirm_box_ok`, `test_droppable`
      - utils -> functions utilized by test_demoqa
      
***Screenshots***
Screenshot 2024-05-06 at 14 21 06

### CODE REVIEW

File: tests/sradu/homework_4/utils.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    elif condition == Condition.ELEMENT_VISIBLE:
+        if locator:
+            return EC.visibility_of_element_located(locator)
+        raise ValueError("Locator is required for ELEMENT_VISIBLE condition.")
+    elif condition == Condition.ELEMENT_CLICKABLE:
+        if locator:
+            return EC.element_to_be_clickable(locator)
+        raise ValueError("Locator is required for ELEMENT_CLICKABLE condition.")
+    else:
+        raise ValueError("Invalid condition specified.")
```
Comment (kudinov-fedor) [inline tests/sradu/homework_4/utils.py:30]: I think it is useless functionality,
EC.alert_is_present - is not worse than enum isnt it?

File: tests/sradu/homework_4/utils.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+
+def wait_for_condition(driver: WebDriver, condition_enum: Condition, locator=None, timeout=DEFAULT_TIMEOUT):
```
Comment (kudinov-fedor) [inline tests/sradu/homework_4/utils.py:33]: I am not sure if this function gives any profit to you , you add additional layer of abstraction but could use selenium functionality directly:
compare:
```
Wait(driver, 10).until(EC.element_to_be_clickable(locator))

wait_for_condition(driver, Condition.ALERT_PRESENT,  locator, timeout)

```

it is the same but you need to rely on someones utill which limits you due to implementation specifics

File: tests/sradu/homework_4/utils.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    for element in elements:
+        if text in element.text:
+            return element
+
+    raise RuntimeError(f"Element with {text} was not found")
+
+
+def click_by_action_chains(driver: WebDriver, element: WebElement):
+    action_chains = ActionChains(driver)
+    action_chains.pause(1).move_to_element(element).pause(1).click().perform()
```
Comment (kudinov-fedor) [inline tests/sradu/homework_4/utils.py:62]: you can get driver from element

element.parent

File: tests/sradu/homework_4/utils.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+
+def wait_for_all_elements(driver, by, value) -> List[WebElement]:
+    return Wait(driver, DEFAULT_TIMEOUT).until(EC.visibility_of_all_elements_located((by, value)))
+
+
+def wait_for_text_to_change(driver, locator, expected_text, timeout=DEFAULT_TIMEOUT):
+    try:
+        element = Wait(driver, timeout).until(EC.visibility_of_element_located(locator))
+        Wait(driver, timeout).until(lambda _: element.text == expected_text)
```
Comment (kudinov-fedor) [inline tests/sradu/homework_4/utils.py:45]: does not EC provide - text to be present method?

File: tests/sradu/homework_4/locators.py [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    LIST_MENU_SUB_ITEM = (By.XPATH, "//div[@class='element-group']//li")
+
+
+class AlertsLocators:
+
+    @classmethod
+    def get_click_btn_for_item(cls, driver: WebDriver, item_text: str) -> WebElement:
+        xpath = (f"//div[@id='javascriptAlertsWrapper']//span[text()='{item_text}']/ancestor::div[contains(@class, "
+                 f"'row')][1]//button")
+        return driver.find_element(By.XPATH, xpath)
```
Comment (kudinov-fedor) [inline tests/sradu/homework_4/locators.py:17]: if it is a locator class it should not interract with driver, and should not provide web element as result, 
this method is inconsistent with any other methods in this module

File: tests/sradu/homework_4/utils.py (hunk 3 above)
Comment (Sofiiar) [reply tests/sradu/homework_4/utils.py:38]: updated to substitute driver to element.parent

File: tests/sradu/homework_4/locators.py [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+    LIST_MENU_ITEM = (By.XPATH, "//div[contains(@class, 'top-card')]")
+    LIST_MENU_SUB_ITEM = (By.XPATH, "//div[@class='element-group']//li")
+
+
+class AlertsLocators:
+
+    @classmethod
+    def get_click_btn_for_item(cls, item_text: str) :
+        xpath = (f"//div[@id='javascriptAlertsWrapper']//span[text()='{item_text}']/ancestor::div[contains(@class, "
+                 f"'row')][1]//button")
```
Comment (kudinov-fedor) [inline tests/sradu/homework_4/locators.py:14]: move to next line in mre logical place e.g.:

```
xpath = (f"//div[@id='javascriptAlertsWrapper']//span[text()='{item_text}']
               f"/ancestor::div[contains(@class, 'row')][1]//button")

```

File: tests/sradu/homework_4/locators.py [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+    def get_locator_for_droppable_text_element(cls, div_id):
+        return By.XPATH, f"//div[@id='{div_id}']//p"
+
+    @classmethod
+    def get_locator_for_draggable_element(cls, div_id):
+        return By.XPATH, f"//div[@id='{div_id}']//div[@id='draggable']"
+
+    @classmethod
+    def get_locator_for_droppable_element(cls, div_id):
+        return By.XPATH, f"//div[@id='{div_id}']//div[@id='droppable']"
```
Comment (kudinov-fedor) [inline tests/sradu/homework_4/locators.py:36]: not sure if you need dynamic locator for one single locator usage in test
is it on purpose?

Comment (Sofiiar) [reply tests/sradu/homework_4/locators.py:36]: Yes, there are hidden block with other draggable and droppable elements. 
I implemented the way if I supposed to do each of them.

File: tests/sradu/homework_4/test_demoqa.py [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+
+    # 4 - пересунути блок `Drag me` до блоку `Drop here`
+    source_locator = DragAndDropLocators.get_locator_for_draggable_element("simpleDropContainer")
+    target_locator = DragAndDropLocators.get_locator_for_droppable_element("simpleDropContainer")
+    source_element = find_element(driver, *source_locator)
+    target_element = find_element(driver, *target_locator)
+    drag_and_drop_by_action_chains(source_element, target_element)
+
+    # 5 - перевірити, що нова назва більшого блоку, в який пересунуто маленький блок, становить `Dropped!`
+    assert wait_for_text_to_change(driver, droppable_text_locator, "Dropped!")
```
Comment (kudinov-fedor) [inline tests/sradu/homework_4/test_demoqa.py:83]: can be replaced by:

```
Wait(driver, DEFAULT_TIMEOUT).until(EC.text_to_be_present_in_element(droppable_text_locator, "Dropped!"))
```

not sure if you need this helper

**Review by kudinov-fedor — APPROVED** [review]

**Review by kudinov-fedor — APPROVED** [review]

**Pull request merged**